auth/auth.go

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package auth
+
+import (
+	"context"
+	"net/http"
+)
+
+type key int
+
+const (
+	// UserIDField is a key
+	UserIDField key = iota
+)
+
+// HTTPAuthService defines how to handle requests for various http authentication methods.
+type HTTPAuthService interface {
+	// Authenticate should return a handler that performs some check on the request coming into the given handler
+	// and then runs the handler if it is. If the request is authenticated successfully, it should call SetUser
+	// to make the callers id available to the service in a global context.
+	Authenticate(next http.Handler) http.Handler
```

Comment:
`NewAuthHandler` might be better? I'd expect a method named `Authenticate` to do the auth work instead of returning a handler.